pkg/bloomtokenizer/tokenizer.go

```diff
@@ -0,0 +1,175 @@
+	t.tokenBuffer = t.tokenBuffer[:0] // Reset the result slice
+	var i int                         // rune index (not position that is measured in the range loop)
+	numToks := 0
+	for _, r := range line {
+
+		// j is the index of the buffer to use
+		for j := 0; j < (t.max - t.min); j++ {
+			// n is the length of the ngram
+			n := j + t.min
+			// pos is the position in the buffer to overwrite
+			pos := i % n
+			t.buffers[j][pos] = r
+
+			if i >= n-1 && (i+1-n)%(t.skip+1) == 0 {
+				t.runeBuffer = reassemble(t.buffers[j], (i+1)%n, t.runeBuffer)
+				if numToks >= cap(t.internalTokenBuffer) || numToks == len(t.internalTokenBuffer) {
+					tok := Token{}
+					tok.Key = make([]byte, 0, 132) // Using a 4 byte token and a chunk identifier, it's really 31 bytes. Adding in for special chars and the like here
+					t.internalTokenBuffer = append(t.internalTokenBuffer, tok)
+				}
+				t.internalTokenBuffer[numToks].Key = t.internalTokenBuffer[numToks].Key[:0]
+				t.internalTokenBuffer[numToks].Key = append(t.internalTokenBuffer[numToks].Key, t.runeBuffer...)
+				t.internalTokenBuffer[numToks].Value = string(t.internalTokenBuffer[numToks].Key)
+				numToks++
+			}
+		}
+		i++
+	}
+	t.tokenBuffer = append(t.tokenBuffer, t.internalTokenBuffer[:numToks]...)
+	return t.tokenBuffer
+}
+
+func reassemble(buf []rune, pos int, result []byte) []byte {
+	result = result[:0] // Reset the result slice
+	for i := 0; i < len(buf); i++ {
+		cur := (pos + i) % len(buf)
+		result = utf8.AppendRune(result, buf[cur])
+	}
+	return result
+}
+
+type WrappedTokenizer struct {
+	t           Tokenizer
+	f           func(Token) Token
+	tokenBuffer []Token
+	prefix      []byte
+	i64buf      []byte
+	i32buf      []byte
+}
+
+func (w *WrappedTokenizer) Tokens(line string) []Token {
+	w.tokenBuffer = w.tokenBuffer[:0] // Reset the result slice
+	toks := w.t.Tokens(line)
+	for _, tok := range toks {
+		w.tokenBuffer = append(w.tokenBuffer, w.f(tok))
+	}
+	return append(w.tokenBuffer, toks...)
+}
+
+func (w *WrappedTokenizer) getSkip() int {
+	return w.t.getSkip()
+}
+
+func (w *WrappedTokenizer) getMin() int {
+	return w.t.getMin()
+}
+
+func (w *WrappedTokenizer) getMax() int {
+	return w.t.getMax()
+}
+
+func ChunkIDTokenizer(t Tokenizer) *WrappedTokenizer {
+	p := make([]byte, 0, 256)
+	return &WrappedTokenizer{
+		t:           t,
+		tokenBuffer: make([]Token, 0, TokenBufferSize),
+		prefix:      p,
+		i64buf:      make([]byte, binary.MaxVarintLen64),
+		i32buf:      make([]byte, 4),
+		f: func(tok Token) Token {
+			tok.Key = append(append(tok.Key, p...), tok.Key...)[len(tok.Key):]
+			tok.Value = string(tok.Key)
+			return tok
+		},
+	}
+}
+
+func (w *WrappedTokenizer) reinit(chk logproto.ChunkRef) {
+	//prefix := fmt.Sprintf("%d:%d:%d:", chk.From, chk.Through, chk.Checksum)
+	w.prefix = w.prefix[:0]
+
+	binary.PutVarint(w.i64buf, int64(chk.From))
+	w.prefix = append(w.prefix, w.i64buf...)
+	w.prefix = append(w.prefix, 58)
```

Comment:
nit: adding the colons to our prefix don't serve a purpose here because they don't help disambiguate keys from each other